query/src/provider/deduplicate/algo.rs

```diff
@@ -0,0 +1,247 @@
+            is_sort_key,
+            ranges,
+        })
+    }
+
+    /// Compute the output record batch that includes the specified ranges
+    fn output_from_ranges(
+        &self,
+        batch: &RecordBatch,
+        dupe_ranges: &DuplicateRanges,
+    ) -> ArrowResult<RecordBatch> {
+        let ranges = &dupe_ranges.ranges;
+
+        // each range is at least 1 large, so any that have more than
+        // 1 are duplicates
+        let num_dupes = ranges.iter().map(|r| r.end - r.start - 1).sum();
+
+        self.num_dupes.add(num_dupes);
+
+        // Special case when no ranges are duplicated (so just emit input as output)
+        if num_dupes == 0 {
+            debug!(num_rows = batch.num_rows(), "No dupes");
+            Self::slice_record_batch(&batch, 0, ranges.len())
+        } else {
+            debug!(num_dupes, num_rows = batch.num_rows(), "dupes");
+
+            // Use take kernel
+            let sort_key_indicies = self.compute_sort_key_indicies(&ranges);
+
+            let take_options = Some(TakeOptions {
+                check_bounds: false,
+            });
+
+            // Form each new column by `take`ing the indices as needed
+            let new_columns = batch
+                .columns()
+                .iter()
+                .enumerate()
+                .map(|(input_index, input_array)| {
+                    if dupe_ranges.is_sort_key[input_index] {
+                        arrow::compute::take(
+                            input_array.as_ref(),
+                            &sort_key_indicies,
+                            take_options.clone(),
+                        )
+                    } else {
+                        // pick the last non null value
+                        let field_indicies = self.compute_field_indicies(&ranges, input_array);
+
+                        arrow::compute::take(
+                            input_array.as_ref(),
+                            &field_indicies,
+                            take_options.clone(),
+                        )
+                    }
+                })
+                .collect::<ArrowResult<Vec<ArrayRef>>>()?;
+            RecordBatch::try_new(batch.schema(), new_columns)
+        }
+    }
+
+    /// Returns an array of indices, one for each input range (which
+    /// index is arbitrary as all the values are the same for the sort
+    /// column in each pk group)
+    ///
+    /// ranges: 0-1, 2-4, 5-6 --> Array[0, 2, 5]
+    fn compute_sort_key_indicies(&self, ranges: &[Range<usize>]) -> UInt64Array {
+        ranges.iter().map(|r| Some(r.start as u64)).collect()
+    }
+
+    /// Returns an array of indices, one for each input range that
+    /// return the first non-null value of `input_array` in that range
+    /// (aka it will pick the index of the field value to use for each
+    /// pk group)
+    ///
+    /// ranges: 0-1, 2-4, 5-6
+    /// input array: A, NULL, NULL, C, NULL, NULL
+    /// --> Array[A, C, NULL]
```

Comment:
The start of this comment and return type indicate this function is returning an array of indices, but the example output in the comment looks like an array of values. Should this line instead be:

```suggestion
    /// --> Array[0, 3, 5]
```

?

Reply:
Yes, this is an astute observation. 👍  Nice catch